indexing-service/src/main/java/org/apache/druid/indexing/common/config/TaskConfig.java

```diff
@@ -300,4 +338,50 @@ private String defaultDir(@Nullable String configParameter, final String default
 
     return configParameter;
   }
+
+  private static class TaskStorageDirTracker
+  {
+    private int taskDirIndex = 0;
+
+    private final List<File> baseTaskDirs = new ArrayList<>();
+
+    private final Map<String, File> taskToTempDirMap = new HashMap<>();
+
+    TaskStorageDirTracker(final List<String> baseTaskDirPaths)
+    {
+      for (String baseTaskDirPath : baseTaskDirPaths) {
+        baseTaskDirs.add(new File(baseTaskDirPath));
+      }
+    }
+
+    List<File> getBaseTaskDirs()
+    {
+      return baseTaskDirs;
+    }
+
+    synchronized File getOrSelectTaskDir(final String taskId)
+    {
+      if (!taskToTempDirMap.containsKey(taskId)) {
+        addTask(taskId, baseTaskDirs.get(taskDirIndex));
+        taskDirIndex = (taskDirIndex + 1) % baseTaskDirs.size();
+      }
+
+      return taskToTempDirMap.get(taskId);
+    }
+
+    synchronized void addTask(final String taskId, final File taskDir)
```

Comment:
please make this private so that it is clear that only way of adding a task to the tracker currently is via `getOrSelectTaskDir`

Reply:
This is needed in the case of restore